Bluetooth: Add notification completed callback

KAGA164 opened the pull request:
The application did not have information about sending a notification for a given attribute.
- A callback was added to confirm the sending of the notification.
- Contextual data has been added to bt_l2cap_send_cb.

jhedberg commented on subsys/net/Kconfig:
```diff
@@ -18,8 +18,8 @@ if NET_BUF
-	default 4
-	range 4 65535 if BT
+	default 8
+	range 8 65535 if BT
```
This will penalize all net_buf users of the system and increase the runtime memory consumption noticeably if the system is configured with many buffers (this will in particular happen when IP networking support is enabled). So I'd suggest you look for a solution which doesn't require increasing the user data. We've been in this kind of situations many times, where it looked like we needed to increase the user data size, but it turned out to always be possible to find an alternative, simple enough solution which didn't require it. If I remember right, this is one of the reasons `struct bt_conn_tx` is independent from any buffer.

joerchan replied:
I'm assuming that the change from 4 to 8 is when you have 64 bits pointers.
Then this line #define conn_tx(buf) ((struct conn_tx_cb *)net_buf_user_data(buf)) requires 8 bytes.
@jhedberg This line was here before, it was only moved. Any suggestions on how this should be handled?

jhedberg replied:
Any update on fixing this user data issue?  It can probably be solved by adding a dedicated user data array along side of acl_tx_pool, in conn.c, and then picking elements from it using `user_data = &acl_tx_user_data[net_buf_id(buf)]`. mesh/adv.c does the same thing since it needs 12 bytes of user data per buffer.

jhedberg replied:
@joerchan typed the above message before seeing yours. Is my suggestion with net_buf_id() clear enough?

jhedberg replied:
@joerchan also, I don't think this has anything to do with 64-bit pointers since Zephyr doesn't support such architectures (right now at least). In the case of this PR it's a function pointer + 4 bytes of user data passed to that function, where the 8 bytes seems to come from.

joerchan replied:
Ok, if there are no 64-bit pointers then 4 bytes of user data should be enough. The net_buf_id(buf) suggestion is clear, thank you.

cvinayak commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -621,11 +622,38 @@ struct notify_data {
+	struct bt_gatt_notify_cb *cb;
+
+	for (cb = callback_list; cb; cb = cb->_next) {
+		if (cb->notify_complete) {
```
Instead of checking null pointer in loop here, please check for null pointers when registering the callback.

cvinayak commented on include/bluetooth/gatt.h:
```diff
@@ -677,6 +677,36 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+	 */
+	void (*notify_complete)(struct bt_conn *conn,
+				const struct bt_gatt_attr *attr);
+	struct bt_gatt_notify_cb *_next;
```
Can we make this the first member in struct (slist node instead) and use slist interfaces for maintaining?

cvinayak commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -833,7 +864,7 @@ int bt_gatt_notify(struct bt_conn *conn, const struct bt_gatt_attr *attr,
 
 	if (conn) {
-		return gatt_notify(conn, attr->handle, data, len);
+		return gatt_notify(conn, attr->handle, attr, data, len);
```
hmmm... do you still need to pass attr->handle when you pass the whole attr?

Vudentz reviewed (changes requested):
Id really like to avoid such callbacks, perhaps make it bt_gatt_notify block if you want to confirm the sending. Btw, on the issue #7659 we didn't conclude what was the real problem which is still not clear to me why the thread is not blocking if it runs out of buffers. Also note that blocking will rate limit the outstanding notification to 1 which can also be archieve by adjusting the number of buffers, so IMO this would not make any different, except if the application is interested in knowing the latency of the stack to dispatch the notifications.

Qbicz commented:
@pdunaj

KAGA164 commented:
@Vudentz I checked the code.  The thread is actually blocked by bt_gatt_notify  if there are no free buffers, but maybe it would be worth adding information about it to the documentation because this function should not be used in interrupts. So it would be assumed that the stack is reliable and every notification will be sent and maybe this callback is not needed. I would like to hear @pdunaj opinion.

pdunaj commented:
Hi, I added use case and some explanation in https://github.com/zephyrproject-rtos/zephyr/issues/7659.

Commit:
subsys: bluetooth: Add notification TX complete callback

This changed added notification complete callback which
gives information if a given notification has been sent.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -74,13 +74,21 @@ struct bt_conn_sco {
 struct bt_conn_tx {
 	sys_snode_t node;
 	bt_conn_tx_cb_t cb;
+	void *context;
```
The convention used elsewhere in the Bluetooth (and Networking) code is to call such context variables that get passed to callbacks user_data.

jhedberg replied:
That said, I kind of like "context" since that's what this is in practice. It's also a shorter name :)

KAGA164 commented:
I changed my approach to this callback. I think that now it is more transparent and simpler.

pdunaj commented on include/bluetooth/gatt.h:
```diff
@@ -699,6 +699,40 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+	/** Characteristic or Characteristic Value attribute. */
+	const struct bt_gatt_attr *attr;
+
+	/** Notification Value callback. */
```
btw - why "Value" or "Attribute"... maybe we should keep it lower case (just like in line 720)?

pdunaj commented:
Hi  @jhedberg , @Vudentz , @cvinayak
Is there anything needed to progress this PR? I would really like to start using it.

jhedberg commented:
I like the new approach better, however I'm a bit uncertain of the necessity of adding the new bt_gatt_notification_params struct. If I understand this right, it's to avoid having too many function parameters? Are you sure it's actually helping reduce stack consumption, since the callers will anyway need to declare the struct on the stack?

jhedberg commented:
@KAGA164 yes, I think I'd prefer if you can get rid of it. In essence the new function is just the old one plus one new parameter, so I think it's actually less confusing to stay similar to the existing function.

jhedberg commented:
Continuing on my earlier thought - I'd even consider going as far as adding a new parameter to the existing function, if it wasn't for the fact that we should try not to break existing APIs.

KAGA164 commented:
@jhedberg Ok I removed this structure. But I left two functions to avoid breaking API.

pdunaj commented:
I was doing some tests with this change. I get the callback but... with one connection interval delay.
i.e. I am able to send 1000 notifications in following time when I set connection interval to:
- 30 us -> 60 s
- 15 us -> 30 s
So it seems that callback comes in next connection interval then data was sent.

Can we have it called just after data is out - on the same connection interval, so I would be able to prepare next notification withing the same time-slice?

pdunaj commented:
Just to be clear - it's better to have this callback then nothing at all but it is not exactly perfect...

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -682,13 +682,14 @@ struct notify_data {
 	int err;
 	u16_t type;
 	const struct bt_gatt_attr *attr;
+	bt_gatt_notify_complete_func_t func;
```
union {attr, func, data, len} with indicate_params?
Or possibly reuse indicate_params, it has the same parameters except for a different typedef for the callback func.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -881,12 +883,14 @@ static u8_t notify_cb(const struct bt_gatt_attr *attr, void *user_data)
 
-int bt_gatt_notify(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-		   const void *data, u16_t len)
+int bt_gatt_notify_with_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
```
+1

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -699,6 +699,29 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+ */
+int bt_gatt_notify_with_cb(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			   const void *data, u16_t len,
+			   bt_gatt_notify_complete_func_t func);
```
If we are going to do this then why not have a _param struct, perhaps it doesn't make sense to have bt_att_req but otherwise it is pretty similar.

jhedberg replied:
You seem to have missed the recent discussion and reasoning about this

jhedberg replied:
@Vudentz note that unlike our other asynchronous callback "param" structs, the way it was used in the previous revision was different: the struct was not persistent (it didn't need to be), rather it was just used to pass function parameters.

pdunaj commented:
I have a chat with @joerchan and sniffed packets transmitted during the connection.
It seems that master is sending 1 packet during a connection event. This way callback is sent on the connection event following one when notification was sent. It ends up with notifications being sent every second connection event.

We could have notification + callback every connection event but this requires that master transmit 2 packets. This may not necessarily be the problem for us although the price is few bytes sent over the link.
Second solution would be to keep 1 additional notification in pipeline: this way data would always be sent callback would keep us in sync with connection interval. Price is that not most recent data is sent.
Third option is that we sync to connection event (avoid callbacks from notifications altogether). This however would require change of the API to add this callback + notifications would have to be non-blocking (we cannot send data if stale packets are still pending as it would cause dragging).

Vudentz commented:
@pdunaj 
> we cannot send data if stale packets are still pending as it would cause dragging

Can you expand what do you mean by dragging? I thought that HID latency is super important otherwise it just build input lag. At least I wouldn't buy anything that just keeping key presses or mouse movements to be sent waiting for a confirmation that has been scheduled over the air, or the idea here is that you would resend? I guess not since ACL is supposed to be reliable if it wasn't we wouldn't be sending stuff without acknowledgement starting with L2CAP.

pdunaj commented:
@Vudentz , sure.

Zephyr has the tx queue where notifications go (there can be more then just HID data put there that make problem worse).
From what I was told when master does not confirm LL packet reception data is re-transmitted. If I hook to connection event and transmit hid data with no confirmation, I can quickly fill up this queue. This will be seen as cursor dragging until user stops moving the device and allow queue to flush itself. This assumes noisy environment. Also making queue shorter will not get rid of the problem really.

I think the best option from 3 described in my previous comment would be to stay with the proposed callback and make master send 2 packets to complete transaction in one connection event.
Second best would be to keep one packet in flight and one ready for transmission.

In short - I don't care about ACK. I need to transfer fresh data with every connection event. I cannot build any pipeline for reason you mentioned in your comment. I don't see zephyr giving any options to allow that.

pdunaj commented:
@jhn-nordic

pdunaj commented:
@Vudentz , @cvinayak , can you please approve if no more comments

Vudentz commented:
@pdunaj 
>Zephyr has the tx queue where notifications go (there can be more then just HID data put there that make problem worse).
From what I was told when master does not confirm LL packet reception data is re-transmitted. If I hook to connection event and transmit hid data with no confirmation, I can quickly fill up this queue. This will be seen as cursor dragging until user stops moving the device and allow queue to flush itself. This assumes noisy environment. Also making queue shorter will not get rid of the problem really.

I guess you are not referring to our conn->tx_queue then, that has no limit on how much you can queue, obviously there is a limit on how much buffers there are available but they can be allocated from different pools. The analogy with the user moving the cursor and that means you build up input lag by queueing them up in the queue, so Bluetooth acts as a rate limiter of the sensor, so it is important that the queue can grow otherwise we would have to start discarding the sensor data.

>I think the best option from 3 described in my previous comment would be to stay with the proposed callback and make master send 2 packets to complete transaction in one connection event.
Second best would be to keep one packet in flight and one ready for transmission.

Im afraid you are mixing layers here, there exists a flow control already between controller and host, it makes sure to look at that before trying to add a new API. For instance BT_CONN_TX_MAX is set to BT_CTLR_TX_BUFFERS:
 https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/Kconfig#L209

> In short - I don't care about ACK. I need to transfer fresh data with every connection event. I cannot build any pipeline for reason you mentioned in your comment. I don't see zephyr giving any options to allow that.

That is up to the controller really, the host stack don't control the scheduling of the packets, it just queue the packets for transmission, there is no discarding of the data either otherwise you would see packet losses when using IPSP and ping.

Anyway I can see the notification callback being useful in case of an isochronous transfer, e.g. audio, since that is normally has a time constraint for each frame, but I haven't seem any GATT profile requiring isochronous transfers instead of just best-effort like everything else in ACL.

Vudentz commented:
@pdunaj @carlescufi just explained me what you meant as follow:
> real-time data that requires the latest "fresh" sample
I couldn't understand this before, that sure requires us to inform when you should produce the next sample, sorry for the inconvenient.

Vudentz commented:
@cvinayak I guess it is your review that is blocking the merging of this now.

pdunaj commented on include/bluetooth/gatt.h:
```diff
@@ -696,8 +718,11 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
  */
-int bt_gatt_notify(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-		   const void *data, u16_t len);
+static inline int bt_gatt_notify(struct bt_conn *conn, const struct bt_gatt_attr *attr,
```
but if you do it you will have copy of this function in every module that includes it

cvinayak replied:
using static declaration in public header forces the definition to be included in each source file. Instead, using inline on the callee; but again frankly, I am not convinced having inline is required, as the compiler optimization will take care to not branch twice when the parameter order is same in the caller and callee (with an additional parameter value NULL).

pdunaj replied:
I don't you will optimize that. Unless you add lto function from where you call has no way of telling that NULL will be added and some other function called.

pdunaj commented:
hi @carlescufi , can it be merged?

carlescufi reviewed (approved):
LGTM. For future reference this is required in order to support "last fresh sample" in HID applications.
